proposals/new-ipfs-gc.md

```diff
@@ -0,0 +1,103 @@
+
+#### Removing Specific Content
+Certain users may only need to remove content that becomes unpinned.  This is particularly true of pinning services such as Pinata, and will be their primary, if not only, method of garbage collection. To remove content that becomes unpinned, the command line workflow is:
+- Unpin the desired content: `ipfs pin rm <cid>`
+- GC the unpinned content: `ipfs repo gc --cid=<cid>`
+
+The above could also be done with a single command, if it adds value: `ipfs pin rm -gc <cid>`
+It may be valuable once local named pinning is available.
+
+In the case that a user wants to remove content that was never pinned, they execute step 2 from above. This tells GC to walk the IPLD DAG identified by the CID and remove any associated blocks that are not referenced.
+
+### Impact
+While impact varies greatly for different types of users, some of our most important users (e.g. Pinata) are heavily impacted and have a need for this feature.  It will have a significant effect on their business operations.  Therefore, this gets an impact rating of 3.
+🔥🔥🔥 
+
+### Leverage
+This fixes a few specific issues, but is not highly leverageable.  The ability to remove specific content may allow implementation of a custom data retention policy.  Therefore this gets a 1 for leverage. 
+🎯
+
+### Confidence
+Medium (level 5) confidence, according to [this scale](https://medium.com/@nimay/inside-product-introduction-to-feature-priority-using-ice-impact-confidence-ease-and-gist-5180434e5b15)_.
+
+## Project definition
+
+### Brief plan of attack
+
+1. Design how reference counting is implemented, exploring what interfaces will change, and what subsystems will be affected.  Address handling storage limits, preventing removal of newly arrived blocks, need for transactions in datastore, etc.
+2. Implement reference counting of blocks in the blockstore.  Counts  are updated by pinner and MFS.
+3. Implement new garbage collector that operates on reference counts instead of pinned CIDs
+4. Implement targeted GC (GC that follows a DAG and deletes associated unreferenced blocks)
+5. Optional, if it adds value: Implement command to remove pin and then immediately GC the unpinned CID
+6. Distribute an experimental version of go-ipfs for evaluation and feedback.
+
+Garbage collection performance issues arise because GC attempts to remove blocks by loading a set of all pinned CIDs into memory and searching for each block’s CID in this set.  The plan attack operates on the following premises:
+
+- Maintaining reference counts for blocks will make it quick to determine if a block can be removed, because it is not necessary to load all pinned CID into memory and search this.
+- Specific content can be removed by performing GC on the blocks associated with a IPLD DAG identified by CID.
+
+### What does done look like?
```

Comment:
Related to other comments about metrics, should done also include metrics about GC time and the time it takes to pin/unpin now given the reference counting?

Reply:
I don't think it makes more sense to have time metrics for pinning, because the pinning process may also need to fetch content from the network as it is being pinned.

I don't think that metrics for unpin become any more important, as this should be equivalent to walking a dag whose.